src/librustc_typeck/check/mod.rs

```diff
@@ -3265,7 +3267,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // an "opportunistic" vtable resolution of any trait bounds on
             // the call. This helps coercions.
             if check_closures {
-                self.select_obligations_where_possible(false);
+                // We don't use `select_obligations_where_possible` to try to figure out if the
+                // obligation is comming from a single fn call argument, and if it is, we point
+                // at the expression corresponding to that argument, instead of the call.
+                if let Err(
+                    mut errors,
+                ) = self.fulfillment_cx.borrow_mut().select_where_possible(self) {
+                    self.point_at_arg_instead_of_call_if_possible(
```

Comment:
Hmm; we could bake this into `fn select_obligations_where_possible` with:
```rust
self.select_obligations_where_possible(false, |errors| {
    self.point_at_arg_instead_of_call_if_possible(
        errors,
        &*final_arg_types,
        sp,
        &args,
    )
});
```

to make things "more automatic" and less open to mistakes.

Some cases will need to pass a `|_| {}` lambda but it also opens up for targeted tweaks in those places too.